File: kangaroo_app_sdk_ios/ios/Classes/features/publicReward/PublicRewardHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class PublicRewardHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_public_reward"

    var eventChannel: String = "customer_sdk/events/get_public_reward"

    func onMethodCall(call: FlutterMethodCall) -> Void? {
        PublicRewardHandler.getPublicReward(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getPublicReward(call: FlutterMethodCall) {
        

        

        

        guard let args = call.arguments else {
            return
        }
        if let myArgs = args as? [String: Any] {
                        guard let rewardSlug = myArgs["rewardSlug"] as? String else {return}

            PublicRewardApi().getPublicReward(
                rewardSlug: rewardSlug
            )
        }
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        PublicRewardApi().publicRewardState
            .serializePublicRewardState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}

File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/publicReward/PublicRewardHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.publicReward

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.publicReward.PublicRewardApi
import features.publicReward.models.PublicRewardModel

import features.publicReward.serializePublicRewardState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class PublicRewardHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_public_reward"

    override val eventChannel: String
        get() = "customer_sdk/events/get_public_reward"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getPublicReward(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getPublicReward(call: MethodCall): String? {
            val result = PublicRewardApi().getPublicReward(
                rewardSlug = call.argument<String>("rewardSlug") as String
            ).toJsonResult<PublicRewardModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        PublicRewardApi().publicRewardState.serializePublicRewardState().observe { result ->
            pushSerializedResultToEventSink<PublicRewardApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}
